src/tags/tests/test_models.py

```diff
@@ -0,0 +1,276 @@
+                badge, _ = self.create_badge(next(self.xp_iter))
+                badge.tags.add(tag_el)
+
+        self.assertEqual(total_xp_by_tags(self.user, tag_list), sum(xp_list))
+
+    def test_multiple_crossing_tags_badges_only(self):
+        """ 
+            See if correct xp is displayed using only multiple tags added to Badge objects
+            each Badge object should have multiple tags
+        """ 
+        tag_list = [(f"tag0-{i}", f"tag1-{i}") for i in range(5)]
+        xp_list = self.xp_list[:5 * len(tag_list)]  # 25
+
+        for tag_set in tag_list:
+            for i in range(len(xp_list) // len(tag_list)):
+                badge, _ = self.create_badge(next(self.xp_iter))
+                badge.tags.add(tag_set[0])
+                badge.tags.add(tag_set[1])
+
+        unpacked_tag_list = [tag_set[index] for tag_set in tag_list for index in range(len(tag_set))]
+        self.assertEqual(total_xp_by_tags(self.user, unpacked_tag_list), sum(xp_list))
+
+    def test_one_tag_multiple_badges_only(self):
+        """ 
+            See if correct xp is displayed using only one tag + badges with multiple badgeassertions
+        """
+        xp_list = self.xp_list[:len(self.set_list)]  # 10
+
+        for i in range(len(xp_list)):
+            badge, ba = self.create_badge(xp_list[i], self.set_list[i])
+            badge.tags.add("TAG")
+
+        self.assertEqual(total_xp_by_tags(self.user, ["TAG"]), sum(xp_list[i] * self.set_list[i] for i in range(len(self.set_list))))
+
+    def test_one_tag_badge_only__different_semesters(self):
+        """ 
+            See if correct xp is displayed using only one tag + badges with multiple badgeassertions in different semesters
+        """ 
+        semester_set = [SiteConfig().get().active_semester] + baker.make("courses.semester", _quantity=2)
+        for semester in semester_set:
+            badge, ba = self.create_badge(100)
+            badge.tags.add("TAG")
+
+            ba[0].semester = semester
+            ba[0].save()
+
+        self.assertEqual(total_xp_by_tags(self.user, ["TAG"]), 100)
+
+    # QUEST + BADGE TESTS
+
+    def test_one_tag_quests_badges(self):
+        """ 
+            See if correct xp is displayed using only 1 tag added to Quest and Badge objects
+        """ 
+        xp_list = self.xp_list[:20]  # 20
+        
+        for i in range(len(xp_list) // 2):
+            quest, _ = self.create_quest(next(self.xp_iter))
+            quest.tags.add("TAG")
+
+            badge, _ = self.create_badge(next(self.xp_iter))
+            badge.tags.add("TAG")
+
+        self.assertEqual(total_xp_by_tags(self.user, ["TAG"]), sum(xp_list))
+
+    def test_multiple_separate_tags_quests_badges(self):
+        """ 
+            See if correct xp is displayed using only multiple tags added to Quest and Badge objects
+            only 1 unique tag per quest and badge
+        """ 
+        tag_list = [f"tag-{i}" for i in range(5 * 2)]  # 10
```

Comment:
Why 10? Why 5x2? Tests are lacking in code comments throughout.

Reply:
10 since 5 per quest and 5 per badge.